### ISSUE TITLE
Problem with bin/chee and local repositories

### ISSUE THREAD

**Issue by arvindsv** [eikek/chee#4]
When building locally using `sbt make-zip`, the generated zip file creates the `chee` script which is supposed to work when placed somewhere in the path. The script generated is this:

```
#!/bin/sh

cd $(dirname $(readlink -e $0))

java -cp lib/fastparse_2.11-0.4.1.jar:lib/scala-library-2.11.8.jar:lib/fastparse-utils_2.11-0.4.1.jar:lib/sourcecode_2.11-0.1.1.jar:lib/better-files_2.11-2.16.0.jar:lib/scala-logging_2.11-3.5.0.jar:lib/logback-classic-1.1.7.jar:lib/logback-core-1.1.7.jar:lib/slf4j-api-1.7.20.jar:lib/scopt_2.11-3.5.0.jar:lib/scrimage-core_2.11-2.1.7.jar:lib/imageio-core-3.2.1.jar:lib/common-lang-3.2.1.jar:lib/common-io-3.2.1.jar:lib/common-image-3.2.1.jar:lib/imageio-jpeg-3.2.1.jar:lib/imageio-metadata-3.2.1.jar:lib/metadata-extractor-2.8.1.jar:lib/xmpcore-5.1.2.jar:lib/commons-io-2.4.jar:lib/pngj-2.1.0.jar:lib/sqlite-jdbc-3.14.2.1.jar:lib/config-1.3.1.jar:lib/bcpg-jdk15on-1.55.jar:lib/bcprov-jdk15on-1.55.jar:lib/mime-util-2.1.3.jar:lib/chee_2.11-0.3.0-SNAPSHOT.jar  $CHEE_OPTS chee.cli.Main "$@"
```

However, when you try to use this with a non-global (erm, local) repository as mentioned in [this post](https://eknet.org/main/introducing_chee__a_photo_managing_tool.html#orgheadline4) - like this - `chee add -r *`, it fails:

```
$ chee add -r *
Error: Path `/path/to/chee-0.3.0-SNAPSHOT/pictures' does not exist!
Try `chee help' for detailed help topics or `--usage' for usage information.
Error: Invalid options
```

If you try it with this: `chee add -r `pwd`/*`, it *works*, but it no longer uses the local `.chee` directory. All of this is happening because of the generated `.chee` script changing the working directory before it starts. When I change the "chee" script to be this, it works:

```
#!/bin/sh

libdir=$(dirname $(readlink -e $0))

java -cp "${libdir}":"${libdir}/lib/fastparse_2.11-0.4.1.jar":"${libdir}/lib/scala-library-2.11.8.jar":"${libdir}/lib/fastparse-utils_2.11-0.4.1.jar":"${libdir}/lib/sourcecode_2.11-0.1.1.jar":"${libdir}/lib/better-files_2.11-2.16.0.jar":"${libdir}/lib/scala-logging_2.11-3.5.0.jar":"${libdir}/lib/logback-classic-1.1.7.jar":"${libdir}/lib/logback-core-1.1.7.jar":"${libdir}/lib/slf4j-api-1.7.20.jar":"${libdir}/lib/scopt_2.11-3.5.0.jar":"${libdir}/lib/scrimage-core_2.11-2.1.7.jar":"${libdir}/lib/imageio-core-3.2.1.jar":"${libdir}/lib/common-lang-3.2.1.jar":"${libdir}/lib/common-io-3.2.1.jar":"${libdir}/lib/common-image-3.2.1.jar":"${libdir}/lib/imageio-jpeg-3.2.1.jar":"${libdir}/lib/imageio-metadata-3.2.1.jar":"${libdir}/lib/metadata-extractor-2.8.1.jar":"${libdir}/lib/xmpcore-5.1.2.jar":"${libdir}/lib/commons-io-2.4.jar":"${libdir}/lib/pngj-2.1.0.jar":"${libdir}/lib/sqlite-jdbc-3.14.2.1.jar":"${libdir}/lib/config-1.3.1.jar":"${libdir}/lib/bcpg-jdk15on-1.55.jar":"${libdir}/lib/bcprov-jdk15on-1.55.jar":"${libdir}/lib/mime-util-2.1.3.jar":"${libdir}/lib/chee_2.11-0.3.0-SNAPSHOT.jar" $CHEE_OPTS chee.cli.Main "$@"
```

With this change, it uses the local repository and the example in the documentation of "add" command works.

The generated file `target/bin/chee` does the right thing. I guess [this line](https://github.com/eikek/chee/blob/979279d9b15a1885b0e0c3143a9e422f98c11b9c/project/ScriptPlugin.scala#L80) needs to be changed.

**Comment by eikek** [comment]
Thank you for reporting! Yes it's really a bad idea to change the working dir if the program uses it to determine things… 😕. And it is really embarassing that all the commands in the blog post won't work…. 

The reason that it works for me is that I usually use the script generated in `target/bin`, I should educate myself to always use the artefacts that I publish.

**Comment by arvindsv** [comment]
:) No worries.

chee is very, very nice and really well documented too. Kudos to you for that!

I was writing my own, in Clojure, using drewnoakes/metadata-extractor as well, before I found chee. I threw that away and am going to use chee for managing my photos. I'm trying to see if it'll work well with a 200GB photo database. Let's see. Thanks for writing this and putting so much thought into it.

**Comment by arvindsv** [comment]
As an aside, the only extra piece of metadata I was using was the [subsecond time](https://github.com/drewnoakes/metadata-extractor/blob/1765d5b9798506600ec438b52c9b135aaded8d87/Source/com/drew/metadata/exif/ExifDirectoryBase.java#L345), which helped me sort burst images a little better sometimes.

**Comment by eikek** [comment]
Thanks for your kind words! I'm very curious about chee coping with your big photo collection! (I currently cannot recall the GB of my photos, but it's below 200G for sure). I guess some issues are waiting… I hope not too many, though. I'd appreciate if you could share any problems you face. 

Also thank you for your side note, this is interesting! I don't do burst images myself, so this just didn't occur to me. I'll create a separate issue for that and start experimenting when I find some time.

By the way, my first try was also using Clojure :-) I wasn't good enough in Clojure (I'm still not) and I had to give up on that.

**Comment by arvindsv** [comment]
@eikek: You'd asked about my experience with the 200GB of images. It has been great so far! It took a while (about 2 hours) to finish that, mostly because it was on a spinning hard drive (one of my redundant backups).

Once it was done, it has been fine and quick:

```
# Shows a total of 27870 images in 9 seconds (very happy!)
$ time chee find --all -p '~:checksum ~#length ~:path~%' >all.photos; echo; wc -l all.photos

real	0m9.521s
user	0m12.581s
sys	0m1.245s

27870 all.photos

# Found 2184 duplicate images in the index
$ cut -f1 -d' ' all.photos  | sort | uniq -d | wc -l
2184

# Shows that it found 189GB of images
$ cut -f2 -d' ' all.photos  | awk '{ sum += $1 } END {print sum / (1024 * 1024 * 1024)}'
189.008
```

There were a couple of small issues. Nothing I couldn't get past easily. I'll investigate a bit more and raise an issue, if I feel it's a bug:

1. I had placed my config in the wrong location (`~/.chee`) so it wasn't picking up the index when I tried to query it. Using `chee info` and `chee config`, I was able to figure that out and fix it. But, the difference in speed was light and day. So, it was obvious to me that the index wasn't being used.

2. `chee add -r` does not like being given a symbolic link to a directory. I had to do a: `chee add -r dir-link/*`. symlinks are useful, because I found that my original idea of multiple `.chee` directories, virtually showing up as one isn't part of the concepts. So, I am using the root repository. This is the one I want to investigate more and see why it's happening. Sorry about bringing it up in this (now closed) issue, but I don't want to lose it. :)

3. After 2 hours of running, it said: `Added: 21973 files; Skipped: 5487 files; in 2:15:17.499`. I wasn't sure what was skipped (I'm sure it's fine). The logs didn't help much, because there weren't 5487 lines of files which were skipped. My suspicion is that the default query and the query I provided both took effect and disqualified 5487 files. I just need to verify that. What is worrying is that: 21973 + 5487 = 27870, which is what the index claims it has. So, it has **skipped** files too. Don't know what that means, at this time.

Overall, I'm very happy! I have plans to organize my images using tags and come up with galleries based on the tags to ensure, visually, that I have the categories right, etc. chee has been very helpful. Thanks!

[EDIT: It actually adds up to 27460. I added 410 files later]

**Comment by eikek** [comment]
@arvindsv thank you a lot for your comments! 
1. That seems strange to me. A `.chee` directory (and so `~/.chee`) is used for “repository mode” and it should pick up `.chee/chee.conf`. But this is not the global config file then, which would be `$HOME/.config/chee.conf`. So it depends where chee is invoked. But if you created a `~/.chee` directory, it should actually pick up the config there if invoked somewhere below `~`.

2. I'm not sure if I understand correctly: I guess multiple `.chee` directories don't work as one, because one would need to query multiple databases and combine the results (not that this is a bad thing, it's just not done). It may very well be, that `add` doesn't like symbolic links (I remember having issues with links several times)–this would be something to fix, though. Don't worry about where to bring something up :)

3. This seems strange to me, too. When using `add`, it would/should only skip files, if the path is already in the index. The query and default query should not do something about it, they select files that are either added or skipped. Files that are removed by the queries don't show up in the results at all. If using a `.chee` repository the path is relative, otherwise an absolute path is used. Hmm. are there maybe links involved that are resolved to equal paths…? The logs should indeed give more a clue about what was happening. At least the stdout should list every path and whether it was skipped or added. But then it is really strange that the index has a total of 27870 if it says 5487 were skipped… I have no idea how that can happen.

Don't hesitate to raise new issues if you think there's something strange. Thanks again for this information.

**Comment by arvindsv** [comment]
Here's what I found (it's not chee's problem):

1. You were right. I tried running `chee find --all` in `/tmp` and that didn't give me anything. That's because it had created a `~/.chee-work/` dir and was using that. I like it using ~/.chee as my root directory everywhere possible. So, I symlinked `~/.chee/chee.conf` to `~/.config/chee/chee.conf`. All my files are now in `~/.chee/` and `chee find` works everywhere using the index!

2. Yes, I somehow thought multiple `.chee` directories would be aggregated into one. That would unnecessarily complicate chee anyway. Instead, I created a directory with symlinks to various locations and used that instead. I am actually going to consolidate into one directory soon. So, I won't even need to do that.

    I'll raise a separate issue for the symlink issue with `chee add`. I have been able to reproduce it with a simple setup.

3. I dug into it a bit more and it was again my mistake. I found that I had done a `chee add` of the inner directory before (probably to verify what you say [here](https://eknet.org/main/projects/chee/manual-0.3.0.html#_locations)). So, when I ran it for the bigger and outer directory, it correctly skipped those files since they were already in the index, just as you said.
